include/windows/osd.h

```diff
@@ -0,0 +1,270 @@
+{
+	unsigned long v = (unsigned long)val;
+	FI_FFSL(v);
+	return 0;
+}
+
+static inline int ffsll(long long val)
+{
+	unsigned long long v = (unsigned long long)val;
+	FI_FFSL(v);
+	return 0;
+}
+
+static inline int asprintf(char** ptr, const char* format, ...)
+{
+	va_list args;
+	va_start(args, format);
+
+	int len = vsnprintf(0, 0, format, args);
+	*ptr = (char*)malloc(len + 1);
+	vsnprintf(*ptr, len + 1, format, args);
+    (*ptr)[len] = 0; /* to be sure that string is enclosed */
+	va_end(args);
+
+	return len;
+}
+
+static inline int gettimeofday(struct timeval* time, struct timezone* zone)
+{
+	const uint64_t shift = 116444736000000000ULL;
+	zone;
+
+	SYSTEMTIME  stime;
+	FILETIME    ftime;
+	uint64_t    utime;
+
+	GetSystemTime(&stime);
+	SystemTimeToFileTime(&stime, &ftime);
+	utime = (((uint64_t)ftime.dwHighDateTime) << 32) + ((uint64_t)ftime.dwLowDateTime);
+
+	time->tv_sec = (long)((utime - shift) / 10000000L);
+	time->tv_usec = (long)(stime.wMilliseconds * 1000);
+	return 0;
+}
+
+static inline char* strsep(char** stringp, const char* delim)
+{
+	char* ptr = *stringp;
+	char* p;
+
+	p = ptr ? strpbrk(ptr, delim) : NULL;
+
+	if(!p)
+		*stringp = NULL;
+	else
+	{
+		*p = 0;
+		*stringp = p + 1;
+	}
+
+	return ptr;
+}
+
+static inline int pthread_join(pthread_t thread, void** exit_code)
+{
+	if (WaitForSingleObject(thread, INFINITE) == WAIT_OBJECT_0)
+	{
+		if (exit_code)
+		{
+			DWORD ex = 0;
+			GetExitCodeThread(thread, &ex);
+			*exit_code = (void*)(uint64_t)ex;
+		}
+		CloseHandle(thread);
+		return 0;
+	}
+
+	return -1;
+}
+
+static int socketpair(int af, int type, int protocol, int socks[2])
+{
+	protocol; /* suppress warning */
+	struct sockaddr_in in_addr;
+	int lsock;
+	int len = sizeof(in_addr);
+
+	if(!socks)
+	{
+		WSASetLastError(WSAEINVAL);
+		return SOCKET_ERROR;
+	}
+
+	socks[0] = socks[1] = (int)INVALID_SOCKET;
+	if ((lsock = socket(af == AF_UNIX ? AF_INET : af, type, 0)) == INVALID_SOCKET)
+	{
+		return SOCKET_ERROR;
+	}
+
+	memset(&in_addr, 0, sizeof(in_addr));
+	in_addr.sin_family = AF_INET;
+	in_addr.sin_addr.s_addr = htonl(0x7f000001);
+
+	if(bind(lsock, (struct sockaddr*)&in_addr, sizeof(in_addr)))
+	{
+		int err = WSAGetLastError();
+		fi_close_fd(lsock);
+		WSASetLastError(err);
+		return SOCKET_ERROR;
+	}
+	if(getsockname(lsock, (struct sockaddr*) &in_addr, &len))
+	{
+		int err = WSAGetLastError();
+		fi_close_fd(lsock);
+		WSASetLastError(err);
+		return SOCKET_ERROR;
+	}
+
+	if (listen(lsock, 1))
+		goto err;
+	if ((socks[0] = WSASocket(af == AF_UNIX ? AF_INET : af, type, 0, NULL, 0, 0)) == INVALID_SOCKET)
+		goto err;
+	if(connect(socks[0], (const struct sockaddr*) &in_addr, sizeof(in_addr)))
+		goto err;
+	if ((socks[1] = accept(lsock, NULL, NULL)) == INVALID_SOCKET)
+		goto err;
+
+	fi_close_fd(lsock);
+	return 0;
+
+	int err;
+err:
+	err = WSAGetLastError();
+	fi_close_fd(lsock);
+	fi_close_fd(socks[0]);
+	fi_close_fd(socks[1]);
+	WSASetLastError(err);
+	return SOCKET_ERROR;
+}
```

Comment:
Please add a windows source file and move most of these functions, especially the longer ones into it and out of header files.